# HG changeset patch
# User Rik <rik@octave.org>
# Date 1387863753 28800
#      Mon Dec 23 21:42:33 2013 -0800
# Node ID d6aaf821cf8f699b98c006de5becb6e8cd37ed24
# Parent  c56ce7c6501949f62db1893977ff51a2290835ef
Add input validation for scatter(3) functions (bug #40898)

* __scatter__.m: Add input validation to scatter family of functions.

diff --git a/scripts/plot/draw/private/__scatter__.m b/scripts/plot/draw/private/__scatter__.m
--- a/scripts/plot/draw/private/__scatter__.m
+++ b/scripts/plot/draw/private/__scatter__.m
@@ -79,16 +79,31 @@ function hg = __scatter__ (varargin)
           && ! (   strcmpi (varargin{istart}, "filled")
                 || strcmpi (varargin{istart}, "fill")))
     c = varargin{istart};
     firstnonnumeric++;
   else
     c = [];
   endif
 
+  ## Validate inputs
+  if (nd == 2 && ! size_equal (x, y)) 
+    error ([fcn ": X and Y must have the same size"]);
+  elseif (nd == 3 && ! size_equal (x, y, z)) 
+    error ([fcn ": X, Y, and Z must have the same size"]);
+  endif
+
+  if (! isscalar (s) && ! size_equal (x, s))
+    error ([fcn ": size of S must match X, Y, and Z"]);
+  endif
+
+  if (rows (c) > 1 && rows (c) != rows (x))
+    error ([fcn ": number of colors in C must match number of points in X"]);
+  endif
+
   newargs = {};
   filled = false;
   have_marker = false;
   marker = "o";
   iarg = firstnonnumeric;
   while (iarg <= nargin)
     arg = varargin{iarg++};
     if (ischar (arg) && (strcmpi (arg, "filled") || strcmpi (arg, "fill")))
